### PR TITLE
depends: Use patch rather than sed in bdb package

### PR DESCRIPTION
**Pull request by nathan-at-least**
This converts a sed command within a makefile into a static patch file in the
`./depends` system for `bdb`. We use `sed` in some dependencies to replace
some "dynamic" stuff, where regular expressions are necessary (such as to
remove build-non-reproducibility, I believe). However, whenever we can use a
patchfile, I feel that's superior because it's easier to manually read a patch
file and it's known that the change is "static".

Equivalent to the first commit in https://github.com/bitcoin/bitcoin/pull/19761

### CODE REVIEW

**Comment by nathan-at-least** [comment]
**tl;dr:** My local test shows the same build artifacts are created, byte-for-byte with and without this PR.

An update on local build comparison:

First show the sources are the two correct revisions:

```
~/tmp/tmp-shell.o0n $ ls srcs/
master  static-patchfile-bdb

~/tmp/tmp-shell.o0n $ for src in srcs/*; do (cd "$src"; git log -1 ; ) ; done
commit 3c8ce84d2455d3cd08e03c67848439e49da4f97c (HEAD -> master, upstream/master, upstream/auto)
Merge: ebc8de390 acf2d0718
Author: Homu <sysadmin@z.cash>
Date:   Mon Oct 12 01:04:01 2020 +0000

    Auto merge of #4777 - daira:fix-bind-warning, r=str4d
    
    Fix a new warning about use of boost::bind placeholders after updating Boost
    
    closes #4774
    
    Signed-off-by: Daira Hopwood <daira@jacaranda.org>
commit 5a47aa5350cb206b82ffdedd9ce8f420bbf36f5f (HEAD -> static-patchfile-bdb, nathan/static-patchfile-bdb)
Author: Nate Wilcox <nathan+dev@electriccoin.co>
Date:   Mon Oct 12 11:57:41 2020 +0900

    Convert a sed command to a static patch file.
```

I have run (basically) `make -C ./depends` in each of those `srcs` directories which are separate git clones. Now let's compare results:

```
~/tmp/tmp-shell.o0n $ sha256sum srcs/*/depends/built/x86*/bdb/*.tar.gz
6f3c81e339051543f148df4a609776a97dfb9e1e9adaecc1f20210a15c230411  srcs/master/depends/built/x86_64-pc-linux-gnu/bdb/bdb-6.2.23-f82ca06e559.tar.gz
149f20b6c59b459ee75b78f8b5d262793fdd1737f8a80ad612d522b4bd39f928  srcs/static-patchfile-bdb/depends/built/x86_64-pc-linux-gnu/bdb/bdb-6.2.23-e9aa81871bc.tar.gz
```

Ok, so the hashes differ, not too surprised. Let's open those up and see how they compare:

```
~/tmp/tmp-shell.o0n $ mkdir built

~/tmp/tmp-shell.o0n $ for src in srcs/*; do branch="$(basename "$src")"; ( mkdir built/$branch; cd built/$branch ; tar -xf ../../$src/depends/built/x86*/bdb/*.tar.gz ; ); done

~/tmp/tmp-shell.o0n $ ls built/*
built/master:
bin  docs  include  lib

built/static-patchfile-bdb:
bin  docs  include  lib

~/tmp/tmp-shell.o0n $ diff -ru built/*

~/tmp/tmp-shell.o0n $ echo $?
0
```

Hm, diff says no difference in file contents! That's a good sign, but I'm a bit surprised. Let's do a sha256sum check on every file:

```
~/tmp/tmp-shell.o0n $ for branch in $(ls built/); do sha256sum $(find ./built/$branch -type f) > $branch.sha256sums; done

~/tmp/tmp-shell.o0n $ head *.sha256sums
==> master.sha256sums <==
8861d1f3db31a239fbd3e1d950e3ea5716aec84f9eccac4267faabf3e1f379d9  ./built/master/docs/api_reference/C/add_data_dir_parameter.html
03949bca9c8c5a214d3de6c7d47ac4830dfce7faea46bd2bb8c0487af968a317  ./built/master/docs/api_reference/C/apiReference.css
6e3c921dbd86645e4a2c2aa223f0490f389ec2e7e9b9fe5869b29b3f5c36e846  ./built/master/docs/api_reference/C/BDB-C_APIReference.pdf
8109b3a2263f91c94dea9c4f966b8792b4ea55a36d5524f13f29d1a894859e99  ./built/master/docs/api_reference/C/blob.html
d0e5d0f188f685dd785b2d941ff5c2811fb2cdf9cf91ee1e00163aa0f4b01ec0  ./built/master/docs/api_reference/C/configuration_reference.html
4c0753bc6959bab8e54cbecd1044f293485da21e4a6c47cf0339f43d40a36964  ./built/master/docs/api_reference/C/db_archive.html
d2b1decfc7f5f39d8527832657dec280ae2050a155bba62554edfd6b2f2eec62  ./built/master/docs/api_reference/C/dbassociate_foreign.html
8852e193bcc983b75b715e73b3195c0625ebbac74b616b13a2b2e5457f4a8e95  ./built/master/docs/api_reference/C/dbassociate.html
9e396a3dfd52bf19a24ca60e8d1d5b3ff3bed7b15807d76db180ee2e21b623b0  ./built/master/docs/api_reference/C/dbcclose.html
4dc13f93bf40731903cfc09e1a145347e74248b59dd041f66a3af36ed0d73912  ./built/master/docs/api_reference/C/dbccmp.html

==> static-patchfile-bdb.sha256sums <==
8861d1f3db31a239fbd3e1d950e3ea5716aec84f9eccac4267faabf3e1f379d9  ./built/static-patchfile-bdb/docs/api_reference/C/add_data_dir_parameter.html
03949bca9c8c5a214d3de6c7d47ac4830dfce7faea46bd2bb8c0487af968a317  ./built/static-patchfile-bdb/docs/api_reference/C/apiReference.css
6e3c921dbd86645e4a2c2aa223f0490f389ec2e7e9b9fe5869b29b3f5c36e846  ./built/static-patchfile-bdb/docs/api_reference/C/BDB-C_APIReference.pdf
8109b3a2263f91c94dea9c4f966b8792b4ea55a36d5524f13f29d1a894859e99  ./built/static-patchfile-bdb/docs/api_reference/C/blob.html
d0e5d0f188f685dd785b2d941ff5c2811fb2cdf9cf91ee1e00163aa0f4b01ec0  ./built/static-patchfile-bdb/docs/api_reference/C/configuration_reference.html
4c0753bc6959bab8e54cbecd1044f293485da21e4a6c47cf0339f43d40a36964  ./built/static-patchfile-bdb/docs/api_reference/C/db_archive.html
d2b1decfc7f5f39d8527832657dec280ae2050a155bba62554edfd6b2f2eec62  ./built/static-patchfile-bdb/docs/api_reference/C/dbassociate_foreign.html
8852e193bcc983b75b715e73b3195c0625ebbac74b616b13a2b2e5457f4a8e95  ./built/static-patchfile-bdb/docs/api_reference/C/dbassociate.html
9e396a3dfd52bf19a24ca60e8d1d5b3ff3bed7b15807d76db180ee2e21b623b0  ./built/static-patchfile-bdb/docs/api_reference/C/dbcclose.html
4dc13f93bf40731903cfc09e1a145347e74248b59dd041f66a3af36ed0d73912  ./built/static-patchfile-bdb/docs/api_reference/C/dbccmp.html
```

At first glance, the hashes seem to match, but the directories are different, so let's normalize those:

```
~/tmp/tmp-shell.o0n $ for branch in $(ls built/); do (cd ./built/$branch ; sha256sum $(find . -type f) > ../../$branch.sha256sums); done               
                                                                                                                                                       
~/tmp/tmp-shell.o0n $ diff -u *.sha256sums                                                                                                             
                                                                                                                                                       
~/tmp/tmp-shell.o0n $ sha256sum *.sha256sums                                                                                                           
59ac8c3587d5d1c44375d01b3a22a42318883bedb75cbfd4f1336931f0bf61b0  master.sha256sums                                                                    
59ac8c3587d5d1c44375d01b3a22a42318883bedb75cbfd4f1336931f0bf61b0  static-patchfile-bdb.sha256sums                                                      
```

It indeed looks like the exact same file contents are produced.

**Review by daira — APPROVED** [review]
utACK

**Comment by str4d** [comment]
This reminds me of an upstream PR I was looking at in the last week. I'll check against that in my review.

**Review by str4d — APPROVED** [review]
utACK. I confirmed that this is equivalent to the change in https://github.com/bitcoin/bitcoin/pull/19761/commits/f36140d00c937b5059cd952d759a77e25330388a.

**Comment by str4d** [comment]
@zkbot r+
